src/Microsoft.Health.Dicom.Core/Features/Query/QueryService.cs

```diff
@@ -58,60 +63,136 @@ public class QueryService : IQueryService
-                var seriesResultMap = seriesComputedResults.ToDictionary<SeriesResult, string>(a => a.StudyInstanceUid + "-" + a.SeriesInstanceUid, StringComparer.OrdinalIgnoreCase);
-                responseMetadata = await Task.WhenAll(queryResult.DicomInstances.Select(x =>
-                GenerateQueryResult(x, seriesResultMap[x.StudyInstanceUid + "-" + x.SeriesInstanceUid].DicomDataset, responseBuilder, getFullMetadata: true, cancellationToken)));
+                if (queryExpression.IELevel == Messages.ResourceType.Study)
+                {
+                    getStudyResponse = true;
+                }
+                else if (queryExpression.IELevel == Messages.ResourceType.Series)
+                {
+                    getSeriesResponse = true;
+                }
             }
         }
-        if (responseMetadata == null)
+
+        // logging to track usage
+        _logger.LogInformation("QueryService result retrieval resultCount:{ResultCount}, studyResultRetrieved:{StudyResultRetrieved}, seriesResultRetrieved:{SeriesResultRetrieved}, fullMetadataRetrieved:{FullMetadataRetrieved}", queryResult.DicomInstances.Count(), getStudyResponse, getSeriesResponse, getFullMetadata);
+
+        // start getting and merging the results based on the source.
+        IEnumerable<DicomDataset> instanceMetadata = null;
+        List<long> versions = queryResult.DicomInstances.Select(i => i.Version).ToList();
+        if (getFullMetadata)
         {
-            IEnumerable<DicomDataset> instanceMetadata = await Task.WhenAll(
+            instanceMetadata = await Task.WhenAll(
                 queryResult.DicomInstances.Select(x => _metadataStore.GetInstanceMetadataAsync(x, cancellationToken)));
-
-            responseMetadata = instanceMetadata.Select(m => responseBuilder.GenerateResponseDataset(m));
         }
+        if (getSeriesResponse)
+        {
+            IReadOnlyCollection<SeriesResult> seriesComputedResults = await _queryStore.GetSeriesResultAsync(partitionKey, versions, cancellationToken);
 
-        return new QueryResourceResponse(responseMetadata, queryExpression.ErroneousTags);
-    }
-
-    // TODO version the API to change the default results to only include indexed columns and control the getFullMetadata param
-    private async Task<DicomDataset> GenerateQueryResult(
-        VersionedInstanceIdentifier versionedInstance,
-        DicomDataset computedMetadata,
-        QueryResponseBuilder queryResponseBuilder,
-        bool getFullMetadata,
-        CancellationToken cancellationToken)
-    {
-        if (getFullMetadata)
+            if (instanceMetadata == null)
+            {
+                instanceMetadata = seriesComputedResults.Select(x => x.DicomDataset);
+            }
+            else
+            {
+                Dictionary<DicomIdentifier, SeriesResult> map = seriesComputedResults.ToDictionary<SeriesResult, DicomIdentifier>(a => new DicomIdentifier(a.StudyInstanceUid, a.SeriesInstanceUid, default));
```

Comment:
Ah, you could also use a simple tuple too like `(string, string)` for the key, since you don't really need the instance part of the identifier here.

Reply:
I will keep the DicomIdentifier for now, considering it has GetHashCode Implemented.